torchtitan/train.py

```diff
@@ -428,14 +487,20 @@ def train(self):
                 sync_every=job_config.fault_tolerance.sync_steps,
             ),
         ):
-            for inputs, labels in self.batch_generator(self.dataloader):
-                if self.step >= job_config.training.steps:
-                    break
+            data_iterator = self.batch_generator(self.dataloader)
+            data_ran_out = False
+            while self.step < job_config.training.steps and not data_ran_out:
                 self.step += 1
                 self.gc_handler.run(self.step)
-                self.train_step(inputs, labels)
+                data_ran_out = self.train_step(data_iterator)
+                if data_ran_out:
```

Comment:
TorchTitan currently doesn't perform force checkpoint if data is depleted. We can fix this but I suggest that we don't do this in this PR.